### PR TITLE
adding nowait clause and guided schedule to the Build functions - to …

### PR DESCRIPTION
**Pull request by RiccardoRossi**
…be tested for efficiency

@roigcarlo could u profile this on MN?

### CODE REVIEW

**Comment by pooyan-dadvand** [comment]
Let me check it on windows
Is it improving the scalability?

**Comment by roigcarlo** [comment]
Not sure yet, I was testing the AMGCL part first, but I'll report the results for this before noon so we can merge it if it improves something

**Comment by roigcarlo** [comment]
Not working for me with ```OMP_NUM_THREADS > 1```. 

The example is big, I uploaded it to dropbox: [Tube_3D](https://www.dropbox.com/s/uqgsup91lkkgg9d/tube_3d_simple.gid.tar?dl=1
)

**Review by roigcarlo — COMMENTED** [review]
Seems to work now and traces look much better! :+1:

**Review by roigcarlo — APPROVED** [review]

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h
Comment (pooyan-dadvand) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: May I ask a (may be naive) question. Why you are using a fix optimum chunk size? and how come the 512?

**Review by pooyan-dadvand — COMMENTED** [review]
Finally is there any improvement?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: the question is not "naive" :-)

according to 
https://software.intel.com/en-us/articles/openmp-loop-scheduling

the 512 is the minimum chunk size. It means it will not go below chunks of 512.
We could put a different number, i did not test it, i saw that it being used like this in an other contest

Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: sorry forgot to answer.
yes improvement is not negligible on 16 cores. 
image

**Comment by RiccardoRossi** [comment]
just to document that i tried with atomic instead of locks and on 4 cores (2+HT) it is around 10%-15% slower. However it is scaling a lot from 2 to 2+HT so i think it would be worth to try out on 16cores

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi I understand that it is slower for 2+HT with atomic. But as you said it scales better from 2 to 4. So it says that 2 is much less efficient?

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand 

i tested two things:
1 - the version with nowait, is faster than the one without nowait. Jordi is trying it out in windows to confirm it.
2 - i also tried, but not uploaded in this branch, to test the code using atomics instead of locks. At least on 2 threads it appears to be slower.

i'll wait for Jordi's feedback before merging

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h
Comment (pooyan-dadvand) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: Very good job!! 👍

**Pull request merged**
